Add non-test matrix compare routine

jwnimmer-tri opened the pull request:
Following up on #4221, make a good name for a matrix comparison check that is more appropriate for production code that the prior `MatrixCompare` helper (which is great for tests).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4240)
<!-- Reviewable:end -->

jwnimmer-tri commented:
+@liangfok for feature review, please.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXH450dPsqu6t33sx7Z:bltds52)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
+@sherm1 for platform review I guess. I like this but am requesting a name change, PTAL.

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXIeE1zxz8MVTrYqsVu:b8hrl9t) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
> ```objc
> /// do not compare as equal elements.
> template <typename DerivedA, typename DerivedB>
> bool is_approx_equal_matrix(const Eigen::MatrixBase<DerivedA>& m1,
> ```

Absolute tolerance is not a robust way to compare matrices in general because it requires knowing the overall scaling and even then only works reliably if all the elements are similarly scaled. I recognize that the uses here were already using absolute tolerance so I think the method is a definite improvement. However, I would like the name to reflect that it is using absolute tolerance, maybe `is_approx_equal_matrix_abs()` or `_abstol()`. That would disambiguate the behavior and also leave room for a fancier `is_approx_equal_matrix_rel()` later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXIihqckVtBWWJo7uD9:bbs8vn1)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXImSOrRHJEtAmNsXPR:bw11klv) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Absolute tolerance is not a robust way to compare matrices in general because it requires knowing the overall scaling and even then only works reliably if all the elements are similarly scaled. I recognize that the uses here were already using absolute tolerance so I think the method is a definite improvement. However, I would like the name to reflect that it is using absolute tolerance, maybe `is_approx_equal_matrix_abs()` or `_abstol()`. That would disambiguate the behavior and also leave room for a fancier `is_approx_equal_matrix_rel()` later.
</details>
Could you look at the three uses in this PR and see if abs comparisons are logically appropriate, or if relative would be better?

Eigen already has built-in relative matrix comparisons, and so the purpose of this PR is to improve upon #4221's attempt to replace a klunky testonly function with an efficient runtime one.

If the cases anyway shouldn't be using abstol in the first place, then I would rather just fix them, instead of writing a common abstol utility that nobody should ever use.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 6 of 8 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/is_approx_equal_matrix.h, line 9 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXImaDDqtgdzMQ-Hq8E:-KXImaDEXUSX2dK--6Rr:bo8t7ni) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L9)):*
> ```objc
> namespace drake {
> 
> /// Return true if and only if the two matrices are equal to within a certain
> ```

"Return" should be "Returns" per style guide.

---

*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXIoz-Kv8Cy_VlQA44y:b-48aw5t) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Could you look at the three uses in this PR and see if abs comparisons are logically appropriate, or if relative would be better?
> 
> Eigen already has built-in relative matrix comparisons, and so the purpose of this PR is to improve upon #4221's attempt to replace a klunky testonly function with an efficient runtime one.
> 
> If the cases anyway shouldn't be using abstol in the first place, then I would rather just fix them, instead of writing a common abstol utility that nobody should ever use.
</details>
Does Eigen have a relative matrix comparison method? Are you referring to `isApprox()`?

http://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#abf727b0aba068172f17c1725cf5d6819

If so, I thought that was absolute?

---

*[drake/common/is_approx_equal_matrix.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXInCJ3YURZDONVltOY:-KXInCJ3YURZDONVltOZ:bqrdbz6) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L17)):*
> ```objc
>                             double tolerance) {
>   return (
>       (m1.rows() == m2.rows()) &&
> ```

BTW, what is the rule w.r.t. when code can be in `.h` files vs. in `.cc` files + explicit instantiation?

---

*[drake/common/test/is_approx_equal_matrix_test.cc, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXInUQc_4ph_7ubC6LP:-KXInUQc_4ph_7ubC6LQ:b-4h0rp3) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/test/is_approx_equal_matrix_test.cc#L6)):*
> ```cpp
> #include "gtest/gtest.h"
> 
> using Eigen::VectorXd;
> ```

Alpha sort.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXIpbqTO67wZKBzUysq:bn3j2bw) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Does Eigen have a relative matrix comparison method? Are you referring to `isApprox()`?
> 
> http://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#abf727b0aba068172f17c1725cf5d6819
> 
> If so, I thought that was absolute?
</details>
Yes, `isApprox`.  See how the precision is scaled by the norm of the vectors under comparison?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/common/test/is_approx_equal_matrix_test.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIpQRlMjEp96QscFXR:-KXIpQRm1pRB699EjCHl:b-ivas1e) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/test/is_approx_equal_matrix_test.cc#L12)):*
> ```cpp
> namespace {
> 
> GTEST_TEST(IsApproxEqualMatrixTest, BasicTest) {
> ```

BTW, given the fact that the comparison is absolute, it might be nice to include in this unit test scenarios where the comparison fails due to the numbers being too small or too large. The difficulty, however, is the threshold at which the failures occur may be system-dependent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/is_approx_equal_matrix.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXInCJ3YURZDONVltOY:-KXIq-VfIKNFM8qGNMpI:b-dsoi6i) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L17)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, what is the rule w.r.t. when code can be in `.h` files vs. in `.cc` files + explicit instantiation?
</details>
For functions in `common` and `math` that operate on Eigen types directly, we've generally been allowing the code to be in headers (and you pretty much have to do it that way).  When we write our own classes that operate on Drake-specific types (e.g., `OutputPort` and `Context`) we tend to prefer code in cc file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/test/is_approx_equal_matrix_test.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIpQRlMjEp96QscFXR:-KXIqrmnK1iZTVl0WeI2:b1qr1jf) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/test/is_approx_equal_matrix_test.cc#L12)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, given the fact that the comparison is absolute, it might be nice to include in this unit test scenarios where the comparison fails due to the numbers being too small or too large. The difficulty, however, is the threshold at which the failures occur may be system-dependent.
</details>
Can you give one specific example (with numbers)?  It seems like two numbers that get large and saturate away their nudge would pass as equal.  Do you mean we'd expect them to show up as unequal?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/test/is_approx_equal_matrix_test.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIpQRlMjEp96QscFXR:-KXIrXre_H6tUBwps-dp:b-kqxhnq) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/test/is_approx_equal_matrix_test.cc#L12)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Can you give one specific example (with numbers)?  It seems like two numbers that get large and saturate away their nudge would pass as equal.  Do you mean we'd expect them to show up as unequal?
</details>
I also don't see how "given the fact that the comparison is absolute" matters?  It seems like the information loss is in the operands, not the comparison?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXIqhpiRRwJ0sALuU3Q:b-xb6x6a) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
Ah yes, now I see that the threshold is scaled. Reading the documentation of `isApprox()`, I'm concerned about the following statement:

> Because of the multiplicativeness of this comparison, one can't use this function to check whether *this is approximately equal to the zero matrix or vector. 

In the three use cases of `is_approx_equal_matrix`, I think it's possible for a zero matrix to be one of the compared matrices.

---

*[drake/common/is_approx_equal_matrix.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXInCJ3YURZDONVltOY:-KXIrybG22eUnvKX4XcT:b-jhyysx) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L17)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> For functions in `common` and `math` that operate on Eigen types directly, we've generally been allowing the code to be in headers (and you pretty much have to do it that way).  When we write our own classes that operate on Drake-specific types (e.g., `OutputPort` and `Context`) we tend to prefer code in cc file.
</details>
Got it. Thanks.

---

*[drake/common/test/is_approx_equal_matrix_test.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIpQRlMjEp96QscFXR:-KXIs5zhMZJElvmxccg5:b-rmm5xj) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/test/is_approx_equal_matrix_test.cc#L12)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I also don't see how "given the fact that the comparison is absolute" matters?  It seems like the information loss is in the operands, not the comparison?
</details>
Actually, I take it back. The problem is given a particular tolerance, there's only a certain range of values that can be compared for equality. Thus, if we know the scale of the values being compared, we can (probably) select a tolerance that works.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXIs9S1cXARS-2XRuLi:b-qb56sx) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Ah yes, now I see that the threshold is scaled. Reading the documentation of `isApprox()`, I'm concerned about the following statement:
> 
> > Because of the multiplicativeness of this comparison, one can't use this function to check whether *this is approximately equal to the zero matrix or vector. 
> 
> In the three use cases of `is_approx_equal_matrix`, I think it's possible for a zero matrix to be one of the compared matrices.
</details>
Eigen's `isApprox()` takes care of the overall scaling but (a) is excessively strict for near-zero matrices, and (b) doesn't handle non-uniform elements well. Consider a diagonally-dominant matrix like [10000 1; 1 10000] compared with [10000 10; 10 10000] to a precision of 1e-3. Under most conditions this should be a failure -- the off diagonal terms are off by a factor of ten (1000% error!). But the Frobenius norm of this matrix is about 14000 so the matrices will be compared with an absolute accuracy of 14, meaning the extreme errors in off-diagonals will slip through. The pure absolute error method proposed here doesn't work well in this case either -- it is likely to require too much accuracy for the diagonals if the off diagonals are tested correctly.

We could do better with an elementwise relative error method that transitions to absolute error for small elements. A workable method is:
```
scaleij = max(|m1(i,j)|, |m2(i,j)|, 0.1) * tol
if (|m1(i,j)-m2(i,j)|) > scalij then FAIL
```
That uses an elementwise relative tolerance tol but shifts to an absolute tolerance of tol/10 for small elements so works for comparisons near zero, assuming those elements are approaching noise. It will not work well for matrices that are scaled to be very tiny but with meaningful tiny values; there is no obvious way to distinguish those from zeroes. 
 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
I made some comments about relative vs. absolute tolerances for the three cases here, and how a reasonably well-behaved matrix checker would work. PTAL.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIxoMCTXWRnpF49hi6:-KXIxoMCTXWRnpF49hi7:ba5aa66) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/multibody/parser_urdf.cc#L115)):*
> ```cpp
>     // is [0, 1].
>     const auto& existing_color = material_iter->second;
>     if (!is_approx_equal_matrix(color_rgba, existing_color, 1e-10)) {
> ```

BTW, colors have unit scaling so the relative and absolute scaling would be about the same. Either would work fine.

---

*[drake/systems/controllers/InstantaneousQPController.cpp, line 586 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIyLjnpGKScBsdqnWB:-KXIyLjnpGKScBsdqnWC:b-5lubb5) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/systems/controllers/InstantaneousQPController.cpp#L586)):*
> ```cpp
> 
>   if (!drake::is_approx_equal_matrix(total_wrench_in_world,
>                                      momentum_rate_of_change, 1e-6)) {
> ```

BTW, here I don't think we know reasonable bounds on the absolute scale of these forces. That may be why the requested tolerance here is so loose. If some of the actual numbers are large, a relative tolerance could be more strict.

---

*[drake/systems/controllers/linear_optimal_control.cc, line 77 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIzTLU-E_hAU-ds3OT:-KXIzTLU-E_hAU-ds3OU:b3bqfm1) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/systems/controllers/linear_optimal_control.cc#L77)):*
> ```cpp
>   DRAKE_DEMAND(R.rows() == m && R.cols() == m);
>   DRAKE_DEMAND(is_approx_equal_matrix(Q, Q.transpose(), 1e-10));
>   DRAKE_DEMAND(is_approx_equal_matrix(R, R.transpose(), 1e-10));
> ```

BTW, if we're accepting approximate symmetry, these tests are likely too strict if either of these matrices contain large entries. If we're expecting near-perfect symmetric matrices, the absolute tol here is too loose but is necessary to avoid failing on the large entries. A relative tolerance could be tighter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXJ-rehnotZQkVSwYdK:b-phcxri)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/controllers/linear_optimal_control.cc, line 77 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIzTLU-E_hAU-ds3OT:-KXMMPp9YGl5NlsnHrd2:b-myvzl0) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/systems/controllers/linear_optimal_control.cc#L77)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, if we're accepting approximate symmetry, these tests are likely too strict if either of these matrices contain large entries. If we're expecting near-perfect symmetric matrices, the absolute tol here is too loose but is necessary to avoid failing on the large entries. A relative tolerance could be tighter.
</details>
it's a good question on what we really should be checking here.  out of curiosity, i just dug into the matlab lqr methods (to the extent that the source is provided) and found that they test for symmetry with a relative tolerance: ```norm(Q-Q',1) > 100*eps*norm(Q,1)```.  But since Q and R are almost always provided as hand-designed constants, it might be reasonable to ask for them to be perfectly symmetric.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Would it make sense to introduce a new macro that only compares matrices when in debug mode and incur zero instruction cost when in non-debug mode?

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXX7EzYaXgM6d8bB2Oz:bcy0kpd)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/controllers/linear_optimal_control.cc, line 77 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIzTLU-E_hAU-ds3OT:-KXcbiqy9t35mx1K2pXO:bb5g4vx) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/systems/controllers/linear_optimal_control.cc#L77)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

it's a good question on what we really should be checking here.  out of curiosity, i just dug into the matlab lqr methods (to the extent that the source is provided) and found that they test for symmetry with a relative tolerance: ```norm(Q-Q',1) > 100*eps*norm(Q,1)```.  But since Q and R are almost always provided as hand-designed constants, it might be reasonable to ask for them to be perfectly symmetric.
</blockquote></details>

@RussTedrake I think you were the original author of these DEMANDs.  Is this PR an improvement vs the CompareMatricies version as originally written.  Should I make any other related improvements to these lines in this PR?

Specifically, this is a behavior change that (1) when NaN or Inf values appear in Q or R, the program fails; and (2) upon an error, the matrix contents are no longer logged to stderr.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/multibody/parser_urdf.cc, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIxoMCTXWRnpF49hi6:-KXc_m5TcKBYwuLYviGV:byvbm6d) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/multibody/parser_urdf.cc#L115)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, colors have unit scaling so the relative and absolute scaling would be about the same. Either would work fine.
</blockquote></details>

I don't think that's true?  Colors near black would have a small norm, at which point relative and absolute become quite different?

@liangfok I wonder why we even have a tolerance here?  Why not demand that the colors are identical?  The colors aren't computed with floating point error, they are directly parsed from a file, I believe?  Is there a reasonable use case for URDFs that have material maps that differ by so little?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/controllers/linear_optimal_control.cc, line 77 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIzTLU-E_hAU-ds3OT:-KXceUnjW6s-2EgJpeJ1:bstlwfy) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/systems/controllers/linear_optimal_control.cc#L77)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@RussTedrake I think you were the original author of these DEMANDs.  Is this PR an improvement vs the CompareMatricies version as originally written.  Should I make any other related improvements to these lines in this PR?

Specifically, this is a behavior change that (1) when NaN or Inf values appear in Q or R, the program fails; and (2) upon an error, the matrix contents are no longer logged to stderr.
</blockquote></details>

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> your changes are only an improvement.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/controllers/InstantaneousQPController.cpp, line 586 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIyLjnpGKScBsdqnWB:-KXce-3Qjs8zgH8-2nj2:bqdwud1) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/systems/controllers/InstantaneousQPController.cpp#L586)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, here I don't think we know reasonable bounds on the absolute scale of these forces. That may be why the requested tolerance here is so loose. If some of the actual numbers are large, a relative tolerance could be more strict.
</blockquote></details>

@sherm1 I agree.

The comparison spelling changed from valueCheck to CompareMatrices in 4a4e1363 "Addressed comments from first round of code review"; prior to that, it got uncommented at a552e7a "fixed mistake in total wrench computation..." using 1e-6 tolerance and was that way for a long time; prior to that, it was originally written in comments in 3dca7cd4 "Fixed size mismatch in computation of xcomdot..." using a tighter **1e-8** tolerance.

@tkoolen Would you like me to take any action related to @sherm1's discussion about tolerances on this sanity check?  For example, I could add a TODO in the code here if you wish?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> Would it make sense to introduce a new macro that only compares matrices when in debug mode and incur zero instruction cost when in non-debug mode?

It seems like that would be unnecessarily obscure vs just putting `DRAKE_ASSERT` around the call to some matrix comparison predicate?

---

Review status: 6 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/common/is_approx_equal_matrix.h, line 9 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXImaDDqtgdzMQ-Hq8E:-KXcZd7BjfUc0x5iuJjU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L9)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

"Return" should be "Returns" per style guide.
</blockquote></details>

Done.

---

*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXch_f3fj-wNOmwTYhM:b-wneir3) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Eigen's `isApprox()` takes care of the overall scaling but (a) is excessively strict for near-zero matrices, and (b) doesn't handle non-uniform elements well. Consider a diagonally-dominant matrix like [10000 1; 1 10000] compared with [10000 10; 10 10000] to a precision of 1e-3. Under most conditions this should be a failure -- the off diagonal terms are off by a factor of ten (1000% error!). But the Frobenius norm of this matrix is about 14000 so the matrices will be compared with an absolute accuracy of 14, meaning the extreme errors in off-diagonals will slip through. The pure absolute error method proposed here doesn't work well in this case either -- it is likely to require too much accuracy for the diagonals if the off diagonals are tested correctly.

We could do better with an elementwise relative error method that transitions to absolute error for small elements. A workable method is:
```
scaleij = max(|m1(i,j)|, |m2(i,j)|, 0.1) * tol
if (|m1(i,j)-m2(i,j)|) > scalij then FAIL
```
That uses an elementwise relative tolerance tol but shifts to an absolute tolerance of tol/10 for small elements so works for comparisons near zero, assuming those elements are approaching noise. It will not work well for matrices that are scaled to be very tiny but with meaningful tiny values; there is no obvious way to distinguish those from zeroes. 
 
</blockquote></details>

FTR I am holding off resolving this until we've iterated on the three call sites to determine what's _actually_ needed here.

---

*[drake/common/test/is_approx_equal_matrix_test.cc, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXInUQc_4ph_7ubC6LP:-KXcZvX9d7BzMnqIDO-B:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/test/is_approx_equal_matrix_test.cc#L6)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Alpha sort.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXbniEe9uK0lim4aART:bb4zkni)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Good point.

---

Review status: 6 of 8 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXck_Z1oqa-FQFN0HjO:bdcv6l8)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIxoMCTXWRnpF49hi6:-KXckmdUCWqdBOYzah9m:b-a1g5bb) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/multibody/parser_urdf.cc#L115)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't think that's true?  Colors near black would have a small norm, at which point relative and absolute become quite different?

@liangfok I wonder why we even have a tolerance here?  Why not demand that the colors are identical?  The colors aren't computed with floating point error, they are directly parsed from a file, I believe?  Is there a reasonable use case for URDFs that have material maps that differ by so little?
</blockquote></details>

I suspect there's error converting the ASCII representation of a `double` to a `double`. It's true, though, that I haven't actually tested to see if this error requires a non-zero tolerance.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIxoMCTXWRnpF49hi6:-KXcnZgFoCAhro3bZznd:bdkeie4) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/multibody/parser_urdf.cc#L115)):*
> I don't think that's true?  Colors near black would have a small norm, at which point relative and absolute become quite different?

This is still unit scaling because the max value is 1. The relative scaling I outlined above would effectively use absolute tolerance for the whole color range.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIxoMCTXWRnpF49hi6:-KXco0GOdMal3sb8COH9:b-kdfk8p) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/multibody/parser_urdf.cc#L115)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

> I don't think that's true?  Colors near black would have a small norm, at which point relative and absolute become quite different?

This is still unit scaling because the max value is 1. The relative scaling I outlined above would effectively use absolute tolerance for the whole color range.
</blockquote></details>

@liangfok But when would that parsing not be deterministic?  It seems to be the same routine called in all cases.

I've removed the tolerance in the latest revision.  If Jenkins passes, I think this discussion could be considered closed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIxoMCTXWRnpF49hi6:-KXcwQDqMJP5ESsQ7ECn:bk0qpm7) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/multibody/parser_urdf.cc#L115)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@liangfok But when would that parsing not be deterministic?  It seems to be the same routine called in all cases.

I've removed the tolerance in the latest revision.  If Jenkins passes, I think this discussion could be considered closed.
</blockquote></details>

SGTM. We can always add back a tolerance if we later find that zero tolerance is too strict.

BTW, instead of using `isApprox()` with a zero tolerance, can we use `==`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIxoMCTXWRnpF49hi6:-KXfD4b0y4JaDB3rOx3c:bz77kne) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/multibody/parser_urdf.cc#L115)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

SGTM. We can always add back a tolerance if we later find that zero tolerance is too strict.

BTW, instead of using `isApprox()` with a zero tolerance, can we use `==`?
</blockquote></details>

Huh, apparently so.  Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

tkoolen commented:
Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/controllers/InstantaneousQPController.cpp, line 586 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIyLjnpGKScBsdqnWB:-KXfKn7NIl9YTkhEuRjO:b-7w1tz9) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/systems/controllers/InstantaneousQPController.cpp#L586)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@sherm1 I agree.

The comparison spelling changed from valueCheck to CompareMatrices in 4a4e1363 "Addressed comments from first round of code review"; prior to that, it got uncommented at a552e7a "fixed mistake in total wrench computation..." using 1e-6 tolerance and was that way for a long time; prior to that, it was originally written in comments in 3dca7cd4 "Fixed size mismatch in computation of xcomdot..." using a tighter **1e-8** tolerance.

@tkoolen Would you like me to take any action related to @sherm1's discussion about tolerances on this sanity check?  For example, I could add a TODO in the code here if you wish?
</blockquote></details>

The computation of `momentum_rate_of_change` is far from trivial, so it's hard to argue precisely about tolerances. I'm of the opinion that time spent coming up with the 'right' tolerance is is not spent productively in this case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 7 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXfWuP_YzqCz4kaXLKy:b-hqqzf4) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FTR I am holding off resolving this until we've iterated on the three call sites to determine what's _actually_ needed here.
</blockquote></details>

@sherm1

So the call sites got resolved:
- Parser was "oops, should be exact anyway", so it no longer relevant.
- QP was "it's not so carefully thought out as all that; leave well enough alone".
- Linear control was "this looks good now".

So we have two call sites, both of which prefer to use absolute tolerance.  The OP defect was "Absolute tolerance is not a good way to do things, so rename the method to clarify to readers what's going on".

Perhaps an overall rename could wait until we offer a method that does something otherwise?  Possibly even this method gains extra arguments, instead of having lots of semi-related names such as: http://stackoverflow.com/questions/15051367/how-to-compare-vectors-approximately-in-eigen

Basically I'm wondering Sherm if you think there is still a defect here, and want to clarify what it is in that case?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Another #4034 flake.
@drake-jenkins-bot linux-gcc-ninja-experimental please

sherm1 commented:
*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXfrUQ8ZitEUl2UsALT:b-p32rvp) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@sherm1

So the call sites got resolved:
- Parser was "oops, should be exact anyway", so it no longer relevant.
- QP was "it's not so carefully thought out as all that; leave well enough alone".
- Linear control was "this looks good now".

So we have two call sites, both of which prefer to use absolute tolerance.  The OP defect was "Absolute tolerance is not a good way to do things, so rename the method to clarify to readers what's going on".

Perhaps an overall rename could wait until we offer a method that does something otherwise?  Possibly even this method gains extra arguments, instead of having lots of semi-related names such as: http://stackoverflow.com/questions/15051367/how-to-compare-vectors-approximately-in-eigen

Basically I'm wondering Sherm if you think there is still a defect here, and want to clarify what it is in that case?
</blockquote></details>

Yes, I still think the unqualified name for a method with this limited functionality is a defect. As evidence: Eigen has `isApprox()` and this is `is_approx_equal_matrix()` which is essentially a longer version of the same name. Both take essentially the same arguments, and one would reasonably expect them to be doing the same thing. But they are substantially different and neither (IMO) is robust enough to serve as a general matrix compare routine. The narrowest fix would be to give this a name that reflects what it does, which is compare two matrices for equality using an absolute tolerance. To keep the generic name, the method should at least use elementwise relative/absolute scaling. I would prefer that but since it apparently isn't needed yet I'm happy to wait.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 7 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXfwN6_JOUJdSdxygYD:b-2xoixi) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Yes, I still think the unqualified name for a method with this limited functionality is a defect. As evidence: Eigen has `isApprox()` and this is `is_approx_equal_matrix()` which is essentially a longer version of the same name. Both take essentially the same arguments, and one would reasonably expect them to be doing the same thing. But they are substantially different and neither (IMO) is robust enough to serve as a general matrix compare routine. The narrowest fix would be to give this a name that reflects what it does, which is compare two matrices for equality using an absolute tolerance. To keep the generic name, the method should at least use elementwise relative/absolute scaling. I would prefer that but since it apparently isn't needed yet I'm happy to wait.
</blockquote></details>

@russtedrake This rework was per your nudge on my PR that introduced the`lpNorm` uses here.  I liked the `lpNorm` spelling because it was unambiguous.  In this discussion, Sherm points out that my choice for a replacement name could be misleading.    Would you like to choose the name of this method for us?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending the re-name of `is_approx_equal_matrix()` to something more specific that indicates it's an absolute comparison of the infinity norms.

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/controllers/InstantaneousQPController.cpp, line 586 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIyLjnpGKScBsdqnWB:-KXgFceylUEgQ7RSesXg:b-o0mi1x) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/systems/controllers/InstantaneousQPController.cpp#L586)):*
<details><summary><i>Previously, tkoolen (Twan Koolen) wrote…</i></summary><blockquote>

The computation of `momentum_rate_of_change` is far from trivial, so it's hard to argue precisely about tolerances. I'm of the opinion that time spent coming up with the 'right' tolerance is is not spent productively in this case.
</blockquote></details>

BTW, perhaps we should just accept the `1e-6` absolute tolerance and revisit in the future when we experimentally determine that it's insufficient.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXgFrYQReyjVst14XZ3:b-7peq8a)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXmOndMFAj_X4_EwSSm:b3brtsx) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@russtedrake This rework was per your nudge on my PR that introduced the`lpNorm` uses here.  I liked the `lpNorm` spelling because it was unambiguous.  In this discussion, Sherm points out that my choice for a replacement name could be misleading.    Would you like to choose the name of this method for us?
</blockquote></details>

Re-ping @russtedrake to resolve this discussion with a course of action, be it a name choice or otherwise?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 8 files at r1, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXn5owk-dpZxkEUd6Vx:b-3aou0n) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Re-ping @russtedrake to resolve this discussion with a course of action, be it a name choice or otherwise?
</blockquote></details>

I'm fine w/ sherm's proposed `is_approx_equal_matrix_abstol`, although I feel like the arguments already specify the type so `is_approx_equal_abstol` would be sufficient for me.  I don't have a strong opinion on this one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
PTAL.

---

Review status: 1 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/is_approx_equal_matrix.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXn7a9_ZTJN9R1TzAse:b-y2egod) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_matrix.h#L13)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

I'm fine w/ sherm's proposed `is_approx_equal_matrix_abstol`, although I feel like the arguments already specify the type so `is_approx_equal_abstol` would be sufficient for me.  I don't have a strong opinion on this one.
</blockquote></details>

Alright, I'll use that one.  Happily, `matrix` and `abstol` have the same number of characters, so formatting stays the same.

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXn9xn7G14XI3lw0wXb:b-83330m)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending CI.

---

Reviewed 9 of 9 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXnB_oitH6RvbpClb4O:bwo9p68)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/is_approx_equal_abstol.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-KXIeE1yjd9yjX-1szOv:-KXnCwsgiYI7YiKOatWP:bo7blmq) ([raw file](https://github.com/robotlocomotion/drake/blob/7fec2dcf8d85869933ad9427b9bfb71efca4b777/drake/common/is_approx_equal_abstol.h#L13)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Alright, I'll use that one.  Happily, `matrix` and `abstol` have the same number of characters, so formatting stays the same.

Done.
</blockquote></details>

LGTM.  And sadly, I can appreciate the humor.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r5, 9 of 9 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4240#-:-KXnrEIosGKywlf4OplK:bnfp4nl)*
<!-- Sent from Reviewable.io -->
